sdlib/dmd/gc.d

```diff
@@ -67,48 +67,42 @@ bool __sd_gc_fetch_alloc_info(void* ptr, void** base, size_t* size,
 	return true;
 }
 
-size_t __sd_gc_get_array_used(void* ptr, PageDescriptor pd) {
-	auto e = pd.extent;
-	if (!e) {
-		return 0;
-	}
-
-	if (pd.isSlab()) {
-		auto si = SlabAllocInfo(pd, ptr);
-		return si.usedCapacity;
-	} else {
-		return e.usedCapacity;
-	}
-}
-
-// TODO: this should only do large blocks, and let the druntime side do slabs.
-bool __sd_gc_set_array_used(void* ptr, PageDescriptor pd, size_t newUsed,
-                            size_t existingUsed) {
+bool __sd_gc_shrink_array_used(void* ptr, size_t newUsed, size_t existingUsed) {
+	auto pd = threadCache.maybeGetPageDescriptor(ptr);
 	auto e = pd.extent;
 	if (!e) {
 		return false;
 	}
 
 	if (pd.isSlab()) {
 		auto si = SlabAllocInfo(pd, ptr);
-		if (existingUsed < size_t.max && existingUsed != si.usedCapacity) {
-			// The existing capacity doesn't match.
+		auto offset = ptr - si.address + 1;
+		if (existingUsed != 0 && si.usedCapacity != existingUsed + offset) {
 			return false;
 		}
 
-		return si.setUsedCapacity(newUsed);
-	} else {
-		if (existingUsed < size_t.max && existingUsed != e.usedCapacity) {
-			// The existing capacity doesn't match.
-			return false;
-		}
+		assert(newUsed + offset <= si.usedCapacity);
+		return si.setUsedCapacity(newUsed + offset);
+	}
 
-		e.setUsedCapacity(newUsed);
+	// Large allocation.
+	auto offset = ptr - e.address + 1;
+	if (existingUsed != 0 && e.usedCapacity != existingUsed + offset) {
+		return false;
 	}
 
+	assert(newUsed + offset <= e.usedCapacity);
+	e.setUsedCapacity(newUsed + offset);
 	return true;
 }
 
+bool __sd_gc_extend_array_used(void* ptr, size_t newUsed, size_t existingUsed) {
+	assert(newUsed > existingUsed);
+	return threadCache
+		.extend(ptr[0 .. existingUsed + 1], newUsed - existingUsed + 1) )
+;
```

Comment:
You have a syntax error here, no?

Reply:
yes. totally forgot to build before pushing. oops.